charms/worker/k8s/src/upgrade.py

```diff
@@ -188,41 +204,34 @@ def _on_upgrade_granted(self, event: UpgradeGrantedEvent) -> None:
             event: The UpgradeGrantedEvent instance.
         """
         with status.context(self.charm.unit, exit_status=ops.ActiveStatus("Ready")):
-            self._upgrade(event)
+            if failure := self._upgrade(event):
+                status.add(failure)
+        self._upgrade_complete = True
 
-    def _upgrade(self, event: Union[ops.EventBase, ops.HookEvent]) -> None:
+    def _upgrade(self, event: Union[ops.EventBase, ops.HookEvent]) -> Optional[ops.StatusBase]:
         """Upgrade the snap workload."""
         trigger = reschedule.PeriodicEvent(self.charm)
-        current_version, _ = snap_version("k8s")
+        current_version, _ = snap_version(SNAP_NAME)
 
         status.add(ops.MaintenanceStatus("Verifying the cluster is ready for an upgrade."))
         if not current_version:
-            log.error("Failed to get the version of the k8s snap.")
-            self.set_unit_failed(cause="Failed to get the version of the k8s snap.")
-            status.add(ops.BlockedStatus("Failed to get the version of the k8s snap."))
-            return
-
-        status.add(ops.MaintenanceStatus("Upgrading the charm."))
+            message = "Failed to get the version of the k8s snap."
+            log.error(message)
+            self.set_unit_failed(cause=message)
+            return ops.BlockedStatus(message)
+
+        if self.charm.lead_control_plane and not self._verify_worker_versions():
+            message = "The k8s worker version requirements are not met."
+            log.error(message)
+            self.set_unit_failed(cause=message)
+            trigger.cancel()
+            return ops.BlockedStatus(message)
 
-        if self.charm.lead_control_plane:
-            if not self._verify_worker_versions():
-                self.set_unit_failed(
-                    cause="The k8s worker charm version does not meet the requirements."
-                )
-                trigger.cancel()
-                return
-
-        self.charm.grant_upgrade()
-        if self.charm.is_control_plane:
-            services = list(K8S_CONTROL_PLANE_SERVICES)
-            bootstrap_datastore = BOOTSTRAP_DATASTORE.get(self.charm)
-            if bootstrap_datastore != DATASTORE_TYPE_K8S_DQLITE:
-                services.remove(K8S_DQLITE_SERVICE)
-            if bootstrap_datastore != DATASTORE_TYPE_ETCD:
-                services.remove(MANAGED_ETCD_SERVICE)
-        else:
-            services = list(K8S_WORKER_SERVICES)
+        snap_services = list_services(SNAP_NAME)
+        services = sorted(name for name, service in snap_services.items() if service["enabled"])
```

Comment:
This is now quite an important change. 

rather than use the charm config to determine which services should be stopped/restarted after the upgrade -- we need to use the current set of snap services on the machine that are running

One reason for this is if we're upgrading a cluster bootstrapped with `k8s-dqlite`  -- on upgrade it could be configured for `managed-etcd`.  We'd fail to stop the dqlite service on upgrade here b/c we're looking at charm config rather than actual running services.   

Let's instead restart the `enabled` services of the k8s snap.